# Практическая работа №1



```
# Выбран кодовый формат
```

В этом блокноте вы потренируетесь работать с основными возможностями, которые предоставляет язык Python, и вспомните темы, которые мы разбирали в модуле:
*   переменные;
*   базовые типы данных: int, float, str, bool;
*   структуры данных: list, tuple, dict, set;
*   функции и методы;
*   форматирование строк;
*   потоки ввода и вывода данных: экран и файл;
*   условный оператор;
*   циклы for и while;
*   обработка исключений.



## Задание 1

Введите с клавиатуры число с плавающей точкой и сохраните его в переменную. Для этого можно воспользоваться функцией input(), в которой аргумент — подсказка для пользователя «Введите радиус:». Именно это значение по смыслу мы будем использовать в следующих заданиях.

In [3]:
r = float(input())

 1.23


Вывести значение переменной, а также тип данных значения, которое в ней хранится, можно при помощи функции print.

In [4]:
print(r, type(r))

1.23 <class 'float'>


Тип данных в переменной отличается от типа float, потому что input всегда возвращает строковый тип данных str. Модифицируйте следующую ячейку так, чтобы тип переменной $r$ был float.

In [5]:
r = float(input("Введите радиус:"))
print(r, type(r))

Введите радиус: 1.23


1.23 <class 'float'>


При правильном применении приведения типов и вводе значения 1.23 вывод ячейки будет следующим:

1.23 <class 'float'>

## Задание 2

Вспомним основные арифметические операции с числами. Выведите результат сложения значения в переменной $r$ и числа $6$, умножения радиуса на два (диаметр), а также квадрата радиуса.

In [6]:
# модифицируйте следующие три строчки, вписав арифметические операции в качестве второго аргумента функции
r = float(input("Введите значение 'r':"))
print("Результат операции r+6:", r+6)
print("Диаметр:", 2*r)
print("Квадрат радиуса:", r**2)

Введите значение 'r': 1.1


Результат операции r+6: 7.1
Диаметр: 2.2
Квадрат радиуса: 1.2100000000000002


При значении $r=1.1$ вывод ячейки должен быть таким:

Результат операции r+6: 7.1

Диаметр: 2.2

Квадрат радиуса: 1.2100000000000002

## Задание 3

Найдите объём шара по формуле: $V = \frac 43 \pi r^3$.

Определим значение константы $\pi$ и радиус шара $r$:

In [10]:
pi = 3.14
r = 1.5

Запишите формулу вычисления объёма, используя арифметические операции в Python:

In [11]:
V = (4/3)*pi*r**3
print("Объем шара:", V)

Объем шара: 14.129999999999999


Чтобы проверить правильность записанного выражения, заметим, что при подстановке $r=1.5$ объём равен $V=14.129999999999999$.

## Задание 4

Одинаковые вычисления удобно выполнять при помощи функций. Определим функцию volume, принимающую единственный аргумент — радиус. Возвращаемое значение функции — найденное значение объёма.

In [14]:
def volume(r):
  pi = 3.14
  V = (4/3)*pi*r**3
  return V

Проверим, что результат работы функции совпал с вычисленным значением. Введите значение радиуса с клавиатуры при помощи функции input(). Затем используйте вызов функции volume внутри функции print.

In [15]:
r = float(input('Введите значение радиуса:'))
print("Объем шара:", volume(r)) # вставьте вызов функции volume после запятой

Введите значение радиуса: 1.5


Объем шара: 14.129999999999999


Убедитесь, что при вводе значения 1.5 вывод функции print в этой ячейке полностью совпадает с выводом в задании 3.

## Задание 5

Представим, что у нас теперь не единственный шар, а целых десять. Радиус первого шарика $r=1$. Радиус второго шарика в 1,5 раза больше первого. И так будем увеличивать радиусы шариков в 1,5 раза, пока не дойдём до последнего. Найдите объёмы всех шаров.

Первым шагом найдём радиусы всех шаров и сохраним их в список.

In [14]:
radius_list = [1]

for i in range(9):
  new_radius = radius_list[-1]*1.5
  radius_list.append(new_radius)

print("Радиусы шаров:", radius_list)
print(type(radius_list[1]))

Радиусы шаров: [1, 1.5, 2.25, 3.375, 5.0625, 7.59375, 11.390625, 17.0859375, 25.62890625, 38.443359375]
<class 'float'>


Радиус первого шара известен, поэтому определим список, в котором это значение уже присутствует. После чего определим цикл, который совершает девять итераций (аргумент функции range). Таким образом мы получим десять значений радиуса.

На каждой итерации мы находим последнее значение списка, подставляя индекс -1, и умножаем его на 1,5, чтобы найти новый радиус. После чего мы должны добавить новое значение в список при помощи метода списков append.

---

Наконец определим список, в котором будут храниться объёмы шаров. Вычислите объёмы внутри цикла при помощи функции, которую написали в задании 4.

In [15]:
volume_list = []
pi = 3.14

for radius in radius_list:
    volume = (4/3)*pi*radius**3
    volume_list.append(volume)

print("Объемы шаров:", volume_list)

Объемы шаров: [4.1866666666666665, 14.129999999999999, 47.68875, 160.94953125, 543.20466796875, 1833.3157543945313, 6187.440671081543, 20882.612264900206, 70478.8163940382, 237866.0053298789]


Вывод ячейки должен содержать:

Объёмы шаров: [4.1866666666666665, 14.129999999999999, 47.68875, 160.94953125, 543.20466796875, 1833.3157543945313, 6187.440671081543, 20882.612264900206, 70478.8163940382, 237866.0053298789]

## Задание 6

Вывод значений в предыдущих заданиях трудночитаемый. В этом задании сделаем вывод красивым и удобным для чтения.

Выведите в каждой строке значение радиуса шара и его объём.

In [16]:
for i in range(10):
  print("Радиус шара: %f. Объем шара: %f" % (radius_list[i], volume_list[i]))

Радиус шара: 1.000000. Объем шара: 4.186667
Радиус шара: 1.500000. Объем шара: 14.130000
Радиус шара: 2.250000. Объем шара: 47.688750
Радиус шара: 3.375000. Объем шара: 160.949531
Радиус шара: 5.062500. Объем шара: 543.204668
Радиус шара: 7.593750. Объем шара: 1833.315754
Радиус шара: 11.390625. Объем шара: 6187.440671
Радиус шара: 17.085938. Объем шара: 20882.612265
Радиус шара: 25.628906. Объем шара: 70478.816394
Радиус шара: 38.443359. Объем шара: 237866.005330


Так выглядит намного лучше! Однако можно сделать ещё красивее, если ограничить количество знаков после запятой (например, до двух) и выделить под каждое значение пространство минимум в десять символов.

In [17]:
for i in range(10):
  print("Радиус шара: %10.2f. Объем шара: %10.2f" % (radius_list[i], volume_list[i]))

Радиус шара:       1.00. Объем шара:       4.19
Радиус шара:       1.50. Объем шара:      14.13
Радиус шара:       2.25. Объем шара:      47.69
Радиус шара:       3.38. Объем шара:     160.95
Радиус шара:       5.06. Объем шара:     543.20
Радиус шара:       7.59. Объем шара:    1833.32
Радиус шара:      11.39. Объем шара:    6187.44
Радиус шара:      17.09. Объем шара:   20882.61
Радиус шара:      25.63. Объем шара:   70478.82
Радиус шара:      38.44. Объем шара:  237866.01


Для форматирования строк можно использовать не только %-форматирование, но и f-строки. Вспомните, как они используются, и добавьте код в следующей ячейке.

In [18]:
print(type(radius_list[1]))
for i in range(10):
    print(f"Радиус шара: {radius_list[i]:10.2f} Объем шара: {volume_list[i]:10.2f}")


<class 'float'>
Радиус шара:       1.00 Объем шара:       4.19
Радиус шара:       1.50 Объем шара:      14.13
Радиус шара:       2.25 Объем шара:      47.69
Радиус шара:       3.38 Объем шара:     160.95
Радиус шара:       5.06 Объем шара:     543.20
Радиус шара:       7.59 Объем шара:    1833.32
Радиус шара:      11.39 Объем шара:    6187.44
Радиус шара:      17.09 Объем шара:   20882.61
Радиус шара:      25.63 Объем шара:   70478.82
Радиус шара:      38.44 Объем шара:  237866.01


Вывод этой ячейки должен полностью совпадать с предыдущей.

## Задание 7

Сохраним вычисленные значения в файл.

Пользуясь менеджером контекста with..as, откроем файл 'write_results.txt' в режиме записи 'w'. Для записи в файл воспользуемся методом write().

In [19]:
with open('write_results.txt', 'w') as file:
  for i in range(10):
    file.write("%.2f %.2f\n" % (radius_list[i], volume_list[i]))

При записи в файл необходимо помнить о символе переноса строки, потому что Python не подставляет его автоматически при использовании метода write(). При форматировании также ограничимся лишь отсечением знаков после точки до второго знака.

Для записи в файл может также использоваться функция print(). Вспомните, какие особенности использования есть у неё при записи в файл, и сохраните данные в новый файл 'print_results.txt'.

In [20]:
with open('print_results.txt', 'w') as file:
  for i in range(10):
    print("%.2f %.2f" % (radius_list[i], volume_list[i]), file=file)

Убедитесь, что содержимое файлов совпадает.

## Задание 8

Потренируемся в чтении файлов. Откроем любой из созданных файлов 'write_results.txt' или 'print_results.txt' в режиме чтения 'r'. Затем сохраним строки файла в список.  

In [21]:
with open('print_results.txt', 'r') as file:
  rows = file.readlines()

print(rows)

['1.00 4.19\n', '1.50 14.13\n', '2.25 47.69\n', '3.38 160.95\n', '5.06 543.20\n', '7.59 1833.32\n', '11.39 6187.44\n', '17.09 20882.61\n', '25.63 70478.82\n', '38.44 237866.01\n']


Полученные строки имеют несколько особенностей. Во-первых, в конце каждого элемента списка есть символ переноса строки '\n'. Во-вторых, каждый элемент списка содержит два значения.

Преобразуйте список следующим образом:
*  при помощи строкового метода strip() удалите символ переноса из конца строки;
*  при помощи метода split() разбейте каждую строку на две подстроки по знаку пробела;
*  сохраните первое значение в список 'first_array', а второе — в список 'second_array';
*  не забудьте преобразовать подстроки в числа, используя явное приведение типов.

In [22]:
first_array = []
second_array = []

for row in rows:
    clean_row = row.strip() # вставьте код для новой строки без символа '\n' в конце
    first_substr, second_substr = clean_row.split() # разделите clean_row на подстроки
    first_array.append(float(first_substr)) # добавьте преобразованное в тип float значение в первый список
    second_array.append(float(second_substr)) # добавьте преобразованное в тип float значение во второй список

print(first_array)
print(second_array)

[1.0, 1.5, 2.25, 3.38, 5.06, 7.59, 11.39, 17.09, 25.63, 38.44]
[4.19, 14.13, 47.69, 160.95, 543.2, 1833.32, 6187.44, 20882.61, 70478.82, 237866.01]


## Задание 9

Воспользуйтесь условным оператором, чтобы определить, в каком массиве (radius_list или first_array) сумма значений больше.

In [23]:
sum_radius_list = sum(radius_list)
sum_first_array = sum(first_array)

if sum_radius_list > sum_first_array:
    print(f'Сумма radius_list больше суммы first_array')
else:
    print(f'Сумма radius_list НЕ больше суммы first_array')
# далее воспользуйтесь условным оператором, чтобы:
# если сумма элементов массива radius_list больше, чем сумма элементов first_array
# вывести "Сумма radius_list больше суммы first_array"
# иначе вывести "Сумма radius_list НЕ больше суммы first_array"

Сумма radius_list больше суммы first_array


Здесь мы воспользовались встроенной функцией sum для нахождения суммы элементов списка.

## Задание 10

Выведите все значения радиусов из первого списка 'first_array', пока сумма объёмов шаров меньше 1000.

Сначала воспользуемся циклом for, в котором на каждой итерации будем обновлять переменную sum_of_volumes (сумма объёмов):

In [ ]:
sum_of_volumes = 0
for i in range(10):
  sum_of_volumes += second_array[i]
  if sum_of_volumes < 1000:
    print("Радиус шара: %f. Сумма объемов: %f" % (first_array[i], sum_of_volumes))
  else:
    break

Радиус шара: 1.000000. Сумма объемов: 4.190000
Радиус шара: 1.500000. Сумма объемов: 18.320000
Радиус шара: 2.250000. Сумма объемов: 66.010000
Радиус шара: 3.380000. Сумма объемов: 226.960000
Радиус шара: 5.060000. Сумма объемов: 770.160000


На каждой итерации цикла мы дополнительно проверяем условие задания. Если оно истинно (сумма объёмов меньше 1000), то выводим необходимый результат. Если оно ложно, то цикл завершает свою работу при помощи оператора break.

Для этой задачи можно воспользоваться циклом while, который уже в своём определении содержит условие. Напишите алгоритм при помощи цикла while, который выполнял бы ту же самую задачу.

In [30]:
sum_of_volumes = 0
i = 0
while sum_of_volumes < 1000:
    i += 1
    sum_of_volumes += second_array[i]
    print("Радиус шара: %f. Сумма объемов: %f" % (first_array[i], sum_of_volumes))


# замените условие (condition) на подходящее
  # вставьте свой код тела цикла здесь

Радиус шара: 1.500000. Сумма объемов: 14.130000
Радиус шара: 2.250000. Сумма объемов: 61.820000
Радиус шара: 3.380000. Сумма объемов: 222.770000
Радиус шара: 5.060000. Сумма объемов: 765.970000
Радиус шара: 7.590000. Сумма объемов: 2599.290000


Цикл while не меняет значение каких-либо переменных автоматически, поэтому необходимо вручную дополнительно определить переменную индекса и увеличивать её на единицу во время каждой итерации.


Если возникают сложности с циклом while, обратитесь к следующему примеру.



Этот цикл while будет запрашивать у пользователя строку и выводить её, пока не встретится строка "exit":

In [24]:
i = 0
while input("Введите строку") != "exit":
  i += 1
  print(f"Строка №{i}")

print("Всего было введено строк:", i)

Введите строку 


Строка №1


Введите строку 


Строка №2


Введите строку 


Строка №3


Введите строку exit


Всего было введено строк: 3


Запустите ячейку и посмотрите, как она работает.